crates/turborepo-lib/src/boundaries.rs

```diff
@@ -0,0 +1,440 @@
+
+            let package_root = self.repo_root().resolve(package_info.package_path());
+
+            let internal_dependencies = self
+                .pkg_dep_graph()
+                .immediate_dependencies(&PackageNode::Workspace(package_name.to_owned()))
+                .unwrap_or_default();
+            let unresolved_external_dependencies =
+                package_info.unresolved_external_dependencies.as_ref();
+
+            let package_diagnostics = self
+                .check_package(
+                    &repo,
+                    &package_root,
+                    &package_info.package_json,
+                    internal_dependencies,
+                    unresolved_external_dependencies,
+                    &source_map,
+                )
+                .await?;
+
+            diagnostics.extend(package_diagnostics);
+        }
+
+        Ok(BoundariesResult {
+            source_map: Arc::new(source_map),
+            diagnostics,
+        })
+    }
+
+    fn is_potential_package_name(import: &str) -> bool {
+        PACKAGE_NAME_REGEX.is_match(import)
+    }
+
+    /// Either returns a list of errors or a single, fatal error
+    async fn check_package(
+        &self,
+        repo: &Option<Mutex<Repository>>,
+        package_root: &AbsoluteSystemPath,
+        package_json: &PackageJson,
+        internal_dependencies: HashSet<&PackageNode>,
+        unresolved_external_dependencies: Option<&BTreeMap<String, String>>,
+        source_map: &SourceMap,
+    ) -> Result<Vec<BoundariesDiagnostic>, Error> {
+        let files = globwalk::globwalk(
+            package_root,
+            &[
+                "**/*.js".parse().unwrap(),
+                "**/*.jsx".parse().unwrap(),
+                "**/*.ts".parse().unwrap(),
+                "**/*.tsx".parse().unwrap(),
+                "**/*.vue".parse().unwrap(),
+                "**/*.svelte".parse().unwrap(),
+            ],
+            &["**/node_modules/**".parse().unwrap()],
+            globwalk::WalkType::Files,
+        )?;
+
+        let mut diagnostics: Vec<BoundariesDiagnostic> = Vec::new();
+        // We assume the tsconfig.json is at the root of the package
+        let tsconfig_path = package_root.join_component("tsconfig.json");
+
+        // TODO: Load tsconfig.json
+        let resolver =
+            Tracer::create_resolver(tsconfig_path.exists().then(|| tsconfig_path.as_ref()));
+
+        for file_path in files {
+            if let Some(repo) = repo {
+                let repo = repo.lock().expect("lock poisoned");
+                if matches!(repo.status_should_ignore(file_path.as_std_path()), Ok(true)) {
+                    continue;
+                }
+            };
+            // Read the file content
+            let Ok(file_content) = tokio::fs::read_to_string(&file_path).await else {
+                return Err(Error::FileNotFound(file_path.to_owned()));
+            };
+
+            let comments = SingleThreadedComments::default();
```

Comment:
there's some boilerplate that is repeated between tracer and boundaries, we could deduplicate but imo it's not critical